File: demo/src/Demo.Middlewares/CorrelationMiddleware.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Serilog.Context;

namespace Demo.Middlewares
{
    public class CorrelationMiddleware
    {
        private const string CorrelationIdHeaderName = "Demo-CorrelationId";
        private readonly RequestDelegate _next;

        public CorrelationMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public async Task InvokeAsync(HttpContext context)
        {
            CorrelationContext.Instance = new CorrelationContext
            {
                CorrelationId = GetCorrelationIdFromRequest(context.Request)
            };

            // Properties are disposable so we can ensure that
            // we remove the property after the next middleware is handled.
            using (LogContext.PushProperty("correlation_id", CorrelationContext.Instance.CorrelationId))
            {
                await _next(context);
            }
        }

        private string GetCorrelationIdFromRequest(HttpRequest request)
        {
            if (request.Headers.Any(header => header.Key == CorrelationIdHeaderName))
            {
                var value = request.Headers[CorrelationIdHeaderName].FirstOrDefault();

                if (!string.IsNullOrWhiteSpace(value))
                {
                    return value.Trim();
                }
            }

            // If there was no header present or its value was empty
            // we can use a new identifier
            return Guid.NewGuid().ToString("D");
        }
    }
}

File: demo/src/Demo.MobileApi/src/correlationmiddleware.ts
import { Request, Response } from "express";
import { Guid } from "./guid";
import bunyan = require("bunyan");
import { NextFunction } from "connect";

export const correlationmiddleware = (req: Request, res: Response, next: NextFunction) => {
    let correlationId = req.headers["Demo-CorrelationId"];
    
    if(correlationId == null || correlationId == "") {
      correlationId = Guid.newGuid();
    }

    let logger = <bunyan>req.app.get("logger");

    req.app.set("logger", logger.child({correlation_id: correlationId}));
    
    next();
  };